pkg/gds/emails/mock.go

```diff
@@ -0,0 +1,124 @@
+package emails
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"net/mail"
+
+	"github.com/sendgrid/rest"
+	sgmail "github.com/sendgrid/sendgrid-go/helpers/mail"
+	"github.com/trisacrypto/directory/pkg/gds/config"
+)
+
+// Returns a mockEmailManager containing a mocked SendGrid client. The mocked SendGrid
+// client validates some of the email metadata, writes the email data to an in-memory
+// structure, and returns a REST response like the real SendGrid does.
+func NewMock(conf config.EmailConfig) (m *mockEmailManager, err error) {
+	m = &mockEmailManager{
+		conf:   conf,
+		Client: &mockSendGridClient{},
+	}
+
+	// Parse the admin and service emails from the configuration
+	if m.serviceEmail, err = mail.ParseAddress(conf.ServiceEmail); err != nil {
+		return nil, fmt.Errorf("could not parse service email %q: %s", conf.ServiceEmail, err)
+	}
+
+	if m.adminsEmail, err = mail.ParseAddress(conf.AdminEmail); err != nil {
+		return nil, fmt.Errorf("could not parse admin email %q: %s", conf.AdminEmail, err)
+	}
+
+	return m, nil
+}
+
+type mockEmailManager struct {
+	conf         config.EmailConfig
+	Client       *mockSendGridClient
+	serviceEmail *mail.Address
+	adminsEmail  *mail.Address
+}
```

Comment:
It's a good start, the trick is to use an interface like you suggest. 

```go
type EmailClient interface {
   Send(email *sgmail.SGMailV3) (*rest.Response, error)
}
```

Then the email manager is defined as:

```go
// EmailManager allows the server to send rich emails using the SendGrid service.
type EmailManager struct {
	conf         config.EmailConfig
	client       EmailClient
	serviceEmail *mail.Address
	adminsEmail  *mail.Address
}
```

At this point, both `mockSendGridClient` and `sendgrid.Client` implement the `EmailClient` interface since they both have the same `Send` method (if there are any other `sendgrid.Client` methods that the `EmailManager` uses, we'll have to add them to the interface and to the `mockSendGridClient` as well). 

Then in the `New` function, we'll use a configuration, e.g. `GDS_EMAIL_TESTING=true` so detect whether or not to create a `mockSendGridClient` in the `EmailManager` or a `sendgrid.Client`. The interface allows us to use either interchangeably. 

Because only the calls to `sendgrid.Client` are mocked, we can test all of the `EmailManager` code without actually executing a SendGrid request.